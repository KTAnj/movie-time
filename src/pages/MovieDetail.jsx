import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link, useParams } from "react-router-dom";
import { fetchMovieById } from "../reducer/slices/movieSlice";
import "../styles/MovieDetail.scss";

const MovieDetail = () => {
  const { id } = useParams();
  const dispatch = useDispatch();

  useEffect(() => {
    if (id) {
      dispatch(fetchMovieById(id));
    }
  }, [id, dispatch]);

  const {
    selectedMovie: movie,
    error,
    isLoading,
    query,
  } = useSelector((state) => state.movies);

  if (isLoading) {
    return <h1>Loading ...</h1>;
  }

  if (!movie) {
    return <h1 className={error && "error"}>{error || "No data found"}</h1>;
  }

  return (
    <>
      {query && (
        <Link className="back-link" to={`/search?query=${query}`}>
          Back to Search Result
        </Link>
      )}
      <div className="row movie-details">
        <div className="col">
          <img
            alt={movie.Title}
            src={
              movie.Poster && movie.Poster !== "N/A"
                ? movie.Poster
                : `${process.env.REACT_APP_HOME_PAGE}/assets/default-movie.jpg`
            }
          />
        </div>
        <div className="d-flex flex-column details-col">
          <h1 className="ml-2">{movie.Title}</h1>
          <small className="ml-2">{movie.Plot}</small>
          <div className="d-flex movie-rating">
            <strong>IMDB Rating</strong>
            <span className="ml-2">{movie.imdbRating}</span>
            <strong className="ml-auto">{`${movie.imdbVotes} Votes`} </strong>
          </div>
          <div className="movie-detail-section">
            <div className="row">
              <strong className="title">Country:</strong>
              <span>{movie.Country}</span>
            </div>
            <div className="row">
              <strong className="title">Released:</strong>
              <span>{movie.Released}</span>
            </div>
            <div className="row">
              <strong className="title">Genre:</strong>
              <span>{movie.Genre}</span>
            </div>
            <div className="row">
              <strong className="title">Director:</strong>
              <span>{movie.Director}</span>
            </div>
            <div className="row">
              <strong className="title">Actors:</strong>
              <span>{movie.Actors}</span>
            </div>
            <div className="row">
              <strong className="title">Awards:</strong>
              <span>{movie.Awards}</span>
            </div>
          </div>
        </div>
      </div>
    </>
  );
};

export default MovieDetail;
