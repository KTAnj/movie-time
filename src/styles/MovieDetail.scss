@import "./variables";
@import "./variables";

.back-link {
  color: $titleColor;
  margin-left: 10px;
}

.movie-details {
  margin-top: 20px;
}

.movie-rating {
  line-height: 2;
  color: $blackColor;
  background: $yelloColor;
  padding: 5px;
  margin: 10px;
}

.movie-detail-section {
  line-height: 2;
  color: $fontColor;
  margin-top: 20px;
  margin-left: 10px;

  .title {
    width: 25%;
  }
}
