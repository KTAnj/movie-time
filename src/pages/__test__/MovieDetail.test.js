import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import MovieDetail from "../MovieDetail";
import * as ReactRouter from "react-router-dom";
import * as redux from "react-redux";

const mockDispatch = jest.fn();

const DUMMY_MOVIE = {
  id: 1,
  Title: "Home Alone 2",
  Poster:
    "https://m.media-amazon.com/images/M/MV5BNDI1MzM0Y2YtYmIyMS00ODE3LTlhZjEtZTUyNmEzMTNhZWU5XkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg",
  year: "2018",
  Plot: "One year after Kevin McCallister was left home alone and had to defeat a pair of bumbling burglars, he accidentally finds himself stranded in New York City - and the same criminals are not far behind.",
  imdbRating: "6",
  imdbVotes: "3,333",
};

const mockState = {
  query: "young",
  selectedMovie: DUMMY_MOVIE,
  error: null,
  isLoading: false,
};

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => mockDispatch,
  useSelector: () => mockState,
}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    id: 1,
  }),
}));

function MockMovieDetail() {
  return (
    <ReactRouter.BrowserRouter>
      <MovieDetail />
    </ReactRouter.BrowserRouter>
  );
}

describe("should render Home component", () => {
  it("shold match movie detail snapshot", () => {
    const tree = renderer.create(<MockMovieDetail />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("render back to search result link", () => {
    render(<MockMovieDetail />);
    const linkElement = screen.getByText("Back to Search Result");
    expect(linkElement).toBeVisible();
  });

  it("render image correctly", () => {
    render(<MockMovieDetail />);
    const imageElement = screen.getByAltText(DUMMY_MOVIE.Title);
    expect(imageElement).toBeVisible();
    expect(imageElement.getAttribute("src")).toContain(DUMMY_MOVIE.Poster);
  });

  it("render default image correctly", () => {
    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({ selectedMovie: { ...DUMMY_MOVIE, Poster: "N/A" } });
    render(<MockMovieDetail />);
    const imageElement = screen.getByAltText(DUMMY_MOVIE.Title);
    expect(imageElement).toBeVisible();
    expect(imageElement.getAttribute("src")).toMatch(/default-movie.jpg$/);
  });

  it("render loading state correctly", async () => {
    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({ isLoading: true });
    render(<MockMovieDetail />);
    const loadingElement = await screen.findByText("Loading ...");
    expect(loadingElement).toBeVisible();
  });

  it("render error state correctly", async () => {
    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({ error: "Server Error" });
    render(<MockMovieDetail />);
    const errorElement = await screen.findByText("Server Error");
    expect(errorElement).toBeVisible();
  });

  it("render no movie state correctly", async () => {
    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({});
    render(<MockMovieDetail />);
    const errorElement = await screen.findByText("No data found");
    expect(errorElement).toBeVisible();
  });
});
