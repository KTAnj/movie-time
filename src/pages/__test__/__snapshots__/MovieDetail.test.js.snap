// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Home component shold match movie detail snapshot 1`] = `
[
  <a
    className="back-link"
    href="/search?query=young"
    onClick={[Function]}
  >
    Back to Search Result
  </a>,
  <div
    className="row movie-details"
  >
    <div
      className="col"
    >
      <img
        alt="Home Alone 2"
        src="https://m.media-amazon.com/images/M/MV5BNDI1MzM0Y2YtYmIyMS00ODE3LTlhZjEtZTUyNmEzMTNhZWU5XkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg"
      />
    </div>
    <div
      className="d-flex flex-column"
    >
      <h1
        className="ml-2"
      >
        Home Alone 2
      </h1>
      <small
        className="ml-2"
      >
        One year after Kevin McCallister was left home alone and had to defeat a pair of bumbling burglars, he accidentally finds himself stranded in New York City - and the same criminals are not far behind.
      </small>
      <div
        className="d-flex movie-rating"
      >
        <strong>
          IMDB Rating
        </strong>
        <span
          className="ml-2"
        >
          6
        </span>
        <strong
          className="ml-auto"
        >
          3,333 Votes
           
        </strong>
      </div>
      <div
        className="movie-detail-section"
      >
        <div
          className="row"
        >
          <strong
            className="title"
          >
            Country:
          </strong>
          <span />
        </div>
        <div
          className="row"
        >
          <strong
            className="title"
          >
            Released:
          </strong>
          <span />
        </div>
        <div
          className="row"
        >
          <strong
            className="title"
          >
            Genre:
          </strong>
          <span />
        </div>
        <div
          className="row"
        >
          <strong
            className="title"
          >
            Director:
          </strong>
          <span />
        </div>
        <div
          className="row"
        >
          <strong
            className="title"
          >
            Actors:
          </strong>
          <span />
        </div>
        <div
          className="row"
        >
          <strong
            className="title"
          >
            Awards:
          </strong>
          <span />
        </div>
      </div>
    </div>
  </div>,
]
`;
